diff --git a/jpeglib.h b/jpeglib.h
index 1a78024..144c777 100644
--- a/jpeglib.h
+++ b/jpeglib.h
@@ -20,6 +20,10 @@
 #ifndef JPEGLIB_H
 #define JPEGLIB_H
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 /*
  * First we include the configuration files that record how this
  * installation of the JPEG library is set up.  jconfig.h can be
@@ -1208,4 +1212,8 @@ struct jpeg_color_quantizer { long dummy; };
 #endif
 #endif
 
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* JPEGLIB_H */
