diff --git a/CMakeLists.txt b/CMakeLists.txt
index 939625d..f3685c7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -691,6 +691,9 @@ if(WITH_TURBOJPEG)
       set_target_properties(turbojpeg PROPERTIES
         LINK_FLAGS "${TJMAPFLAG}${TJMAPFILE}")
     endif()
+    if(UNIX)
+      target_link_libraries(turbojpeg m)
+    endif()
 
     add_executable(tjunittest tjunittest.c tjutil.c md5/md5.c md5/md5hl.c)
     target_link_libraries(tjunittest turbojpeg)
